docs/GettingStartedAriesDev/PlugIns.md

```diff
@@ -0,0 +1,172 @@
+* `--block-plugin` - identifies plug-ins (includign built-ins) that are *not* to be loaded
+* `--plugin-config` - identify a configuration parameter for a plug-in
+* `--plugin-config-value` - identify a *value* for a plug-in configuration
+
+
+The `--plug-in` parameter specifies a package that is loaded by Aca-Py at runtime, and extends Aca-Py by adding support for additional protocols and message types, and/or extending the Admin API with additional endpoints.
+
+The original plug-in design (which we will call the "old" model) explicitely indluded `message_types.py` `routes.py` (to add Admin API's).  But functionality was added later (we'll call this the "new" model) to allow the plug-in to include a generic `setup` package that could perform arbitrary initialization.  The "new" model also includes support for a `definition.py` file that can specify plug-in version information  (major/minor plug-in version, as well as the minimum supported version (if another agent is running an older version of the plug-in)).
+
+### setup.py
+
+If a setup method is provided, it will be called.  If not, the `message_types.py` and `routes.py` will be explicitely loaded.
+
+TODO I couldn't find an implementation of a custom `setup` in any of the existing plug-ins, so I'm not completly sure what are the best practices for this option.
+
+### message_types.py
+
+When loading a plug-in, if there is a `message_types.py` available, Aca-Py will check the following attributes to initialize the protocol(s):
+
+- `MESSAGE_TYPES` - identifies message types supported by the protocol
+- `CONTRROLLERS` - identifies protocol controllers
+
+### routes.py
+
+If `routes.py` is available, then Aca-Py will call the following functions to initialize the Admin endpoints:
+
+- `register()` - registers routes for the new Admin endpoints
+- `register_events()` - registers an events this package will listen for/respond to
+
+### definition.py
+
+If `definition.py` is available, Aca-Py will read this package to determine protocol version information.  An example follows (this is an example that specifies two protocol versions):
+
+```
+versions = [
+    {
+        "major_version": 1,
+        "minimum_minor_version": 0,
+        "current_minor_version": 0,
+        "path": "v1_0",
+    },
+    {
+        "major_version": 2,
+        "minimum_minor_version": 0,
+        "current_minor_version": 0,
+        "path": "v2_0",
+    },
+]
+```
+
+The attributes are:
+
+- `major_version` - specifies the protocol major version
+- `current_minor_version` - specifies the protocol minor version
+- `minimum_minor_version` - specifies the minimum supported version (if a lower version is installed in another agent)
+- `path` - specifies the sub-path within the package for this version
+
+
+## Loading Aca-Py Plug-Ins at Runtime
+
+The load sequence for a plug-in (the "Startup" class depends on how Aca-Py is running - `upgrade`, `provision` or `start`):
+
+```mermaid
```

Comment:
wow - this section is fantastic. this would have saved me days of jumping in and out of code and poking around.